### PR TITLE
Implement: attach + tests, re-atttach + tests, simple timer

### PR DESCRIPTION
**Pull request by mik155**

### CODE REVIEW

**Comment by rvbnsk** [comment]
General comment before review. Please turn on lint from github workflows. It should be present in PR

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 1]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
-                                userPortMock,
-                                timerPortMock};
-};
-
-struct ApplicationNotConnectedTestSuite : ApplicationTestSuite
-{};
-
-TEST_F(ApplicationNotConnectedTestSuite, todo)
-{
-}
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:37]: Probably you should fix indentation. "class" should not be indented and it was OK before

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 2]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    class ApplicationTestSuite : public Test
+    {
+    protected:
+        const common::PhoneNumber PHONE_NUMBER{112};
+        const common::BtsId btsId{333};
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:20]: keep c++11 brace init formatting style:
```c++
const common::PhoneNumber PHONE_NUMBER{ 112 };
const common::BtsId btsId{ 333 };
```

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 3]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    {
+        void sendAttachRequestOnSib();
+    };
+
+    void ApplicationNotConnectedTestSuite::sendAttachRequestOnSib()
+    {
+        using Duration = std::chrono::milliseconds;
+        EXPECT_CALL(btsPortMock, sendAttachRequest(btsId));
+        EXPECT_CALL(userPortMock, showConnecting());
+        EXPECT_CALL(timerPortMock, startTimer(Duration(500)));
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:47]: magic number, exctract into variable

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 4]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+        NiceMock<common::ILoggerMock> loggerMock;
+        StrictMock<IBtsPortMock> btsPortMock;
+        StrictMock<IUserPortMock> userPortMock;
+        StrictMock<ITimerPortMock> timerPortMock;
+
+        Application objectUnderTest{PHONE_NUMBER,
+                                    loggerMock,
+                                    btsPortMock,
+                                    userPortMock,
+                                    timerPortMock};
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:31]: as comment above

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 5]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    TEST_F(ApplicationConnectingTestSuite, shallConnectOnAttachAccept)
+    {
+        connectOnAttachAccept();
+    }
+
+    TEST_F(ApplicationConnectingTestSuite, shallNotConnectOnAttachReject)
+    {
+           EXPECT_CALL(userPortMock, showNotConnected());
+           EXPECT_CALL(timerPortMock, stopTimer());
+           objectUnderTest.handleAttachReject();
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:85]: extra indentation, fix please

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 6]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    }
+
+    //RE-ATTACH
+    TEST_F(ApplicationConnectingTestSuite, shallNotConnectOnConnectionDrop)
+    {
+           EXPECT_CALL(userPortMock, showNotConnected());
+           EXPECT_CALL(timerPortMock, stopTimer());
+           objectUnderTest.handleBTSDisconnected();
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:93]: extra indentation

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 7]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    ApplicationConnectedTestSuite::ApplicationConnectedTestSuite()
+    {
+        connectOnAttachAccept();
+    }
+
+    //RE-ATTACH
+    TEST_F(ApplicationConnectedTestSuite, shallDisconnectOnConnectionDrop)
+    {
+           EXPECT_CALL(userPortMock, showNotConnected());
+           objectUnderTest.handleBTSDisconnected();
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:110]: extra indentation

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 8]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
+    }
+
+
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:113]: blank lines, remove

Comment (rvbnsk) [reply src/UE/Tests/Application/ApplicationTestSuite.cpp:113]: only one extra blank line should be present before namespace ending

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 9]
```diff
@@ -13,27 +13,103 @@ namespace ue
...
 
 }
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:115]:
```c++
} // namespace ue
```
For better readability keep informing when namespace ends

File: src/UE/Tests/Application/Ports/BtsPortTestSuite.cpp [hunk 10]
```diff
@@ -22,13 +22,17 @@ class BtsPortTestSuite : public Test
     StrictMock<IBtsEventsHandlerMock> handlerMock;
     StrictMock<common::ITransportMock> transportMock;
     common::ITransport::MessageCallback messageCallback;
+    common::ITransport::DisconnectedCallback disconnectedCallback;
+
```
Comment (rvbnsk) [inline src/UE/Tests/Application/Ports/BtsPortTestSuite.cpp:26]: extra blank line, remove

File: src/UE/Tests/Application/Ports/BtsPortTestSuite.cpp [hunk 11]
```diff
@@ -41,6 +45,7 @@ class BtsPortTestSuite : public Test
 
 TEST_F(BtsPortTestSuite, shallRegisterHandlersBetweenStartStop)
 {
+
```
Comment (rvbnsk) [inline src/UE/Tests/Application/Ports/BtsPortTestSuite.cpp:48]: is this a kind of letfover? It should be left like this?

**Review by rvbnsk — CHANGES_REQUESTED** [review]
Generally OK, please fix minor things

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 12]
```diff
@@ -13,103 +13,101 @@ namespace ue
...
+}
+
+TEST_F(ApplicationConnectingTestSuite, shallNotConnectOnAttachReject)
+{
+    EXPECT_CALL(userPortMock, showNotConnected());
+    EXPECT_CALL(timerPortMock, stopTimer());
+    objectUnderTest.handleAttachReject();
+}
+
+//RE-ATTACH
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:88]: comment as in standard, eg:
```cpp
// RE-ATTACH
```
Let's keep the code clean 😉

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 13]
```diff
@@ -13,103 +13,101 @@ namespace ue
...
+TEST_F(ApplicationConnectingTestSuite, shallNotConnectOnConnectionDrop)
+{
+    EXPECT_CALL(userPortMock, showNotConnected());
+    EXPECT_CALL(timerPortMock, stopTimer());
+    objectUnderTest.handleBTSDisconnected();
+}
+//********************************************************************************************
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:95]: remove

File: src/UE/Application/Ports/TimerPort.cpp [hunk 14]
```diff
@@ -22,11 +22,22 @@ void TimerPort::stop()
...
 
+void TimerPort::run(Duration duration)
+{
+    std::unique_lock<std::mutex> cond_lock(mtx_stop);
+    if(isStopped.wait_for(cond_lock, duration) == std::cv_status::timeout)
+        handler->handleTimeout();
+}
+
+
 }
```
Comment (rvbnsk) [inline src/UE/Application/Ports/TimerPort.cpp:43]: comment end of namespace

File: src/UE/Application/Ports/TimerPort.hpp [hunk 15]
```diff
@@ -1,5 +1,8 @@
 #pragma once
 
+#include <future>
+#include <thread>
```
Comment (rvbnsk) [inline src/UE/Application/Ports/TimerPort.hpp:4]: what is this includes for?
For std::mutex you could use
```cpp
#include <mutex>
```
for condition_variable use:
```cpp
#include <condition_variable>
```
I think future is not needed here

**Review by rvbnsk — CHANGES_REQUESTED** [review]
Sory for waiting for a review for a long time 😉

File: src/UE/Application/Ports/TimerPort.cpp [hunk 16]
```diff
@@ -22,11 +22,22 @@ void TimerPort::stop()
...
 {
     logger.logDebug("Stop timer");
+    isStopped.notify_all();
 }
 
+void TimerPort::run(Duration duration)
+{
+    std::unique_lock<std::mutex> cond_lock(mtx_stop);
+    if(isStopped.wait_for(cond_lock, duration) == std::cv_status::timeout)
+        handler->handleTimeout();
```
[the remark this answers is not in the document]
Comment (mik155) [reply src/UE/Application/Ports/TimerPort.cpp]: - Could cond_lock be const? - compiler doesn't like that

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 17]
```diff
@@ -16,24 +16,95 @@ using namespace ::testing;
...
 
 struct ApplicationNotConnectedTestSuite : ApplicationTestSuite
-{};
+{
+    void sendAttachRequestOnSib();
+};
+
+void ApplicationNotConnectedTestSuite::sendAttachRequestOnSib()
+{
+    const long bts_response_time_ms = 500;
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:41]: constexpr

File: src/UE/Tests/Application/ApplicationTestSuite.cpp [hunk 18]
```diff
@@ -16,24 +16,95 @@ using namespace ::testing;
...
+
+    EXPECT_CALL(btsPortMock, sendAttachRequest(btsId));
+    EXPECT_CALL(userPortMock, showConnecting());
+    EXPECT_CALL(timerPortMock, startTimer(ITimerPort::Duration(bts_response_time_ms)));
+    objectUnderTest.handleSib(btsId);
+
```
Comment (rvbnsk) [inline src/UE/Tests/Application/ApplicationTestSuite.cpp:47]: remove blank line

**Review by rvbnsk — CHANGES_REQUESTED** [review]

**Review by rvbnsk — APPROVED** [review]
lgtm

**Review by rvbnsk — APPROVED** [review]

**Review by szwos — APPROVED** [review]

**Pull request merged**